internal/blockchain/validate.go

```diff
@@ -1014,6 +1035,160 @@ func (b *BlockChain) isOldBlockVersionByMajority(header *wire.BlockHeader, block
 	return b.isMajorityVersion(nextVersion, prevNode, rejectNumRequired)
 }
 
+// checkDifficultyPositional ensures the difficulty specified in the block
+// header matches the calculated difficulty based on the difficulty retarget
+// rules.  These checks do not, and must not, rely on having the full block data
+// of all ancestors available.
+//
+// This function is safe for concurrent access.
+func (b *BlockChain) checkDifficultyPositional(header *wire.BlockHeader, prevNode *blockNode) error {
+	// -------------------------------------------------------------------------
+	// The ability to determine whether or not the blake3 proof of work agenda
+	// is active is not possible in the general case here because that relies on
+	// additional context that is not available in the positional checks.
+	// However, it is important to check for valid bits in the positional checks
+	// to protect against various forms of malicious behavior.
+	//
+	// Thus, with the exception of the special cases where it is possible to
+	// definitively determine the agenda is active, allow valid difficulty bits
+	// under both difficulty algorithms while rejecting blocks that satisify
+	// neither here in the positional checks and allow the contextual checks
+	// that happen later to ensure the difficulty bits are valid specifically
+	// for the correct difficulty algorithm as determined by the state of the
+	// blake3 proof of work agenda.
+	// -------------------------------------------------------------------------
+
+	// Ensure the difficulty specified in the block header matches the
+	// calculated difficulty using the algorithm defined in DCP0011 when the
+	// blake3 proof of work agenda is always active.
+	if b.isBlake3PowAgendaForcedActive() {
+		blake3Diff := b.calcNextBlake3Diff(prevNode)
+		if header.Bits != blake3Diff {
+			str := fmt.Sprintf("block difficulty of %d is not the expected "+
+				"value of %d", header.Bits, blake3Diff)
```

Comment:
we may want to change all of the formatted errors in this function to describe which hashing algorithm was used to improve debuggability.

Reply:
sounds good to me

Reply:
Updated.